src/main/java/ch/njol/skript/lang/function/DefaultFunction.java

```diff
@@ -0,0 +1,341 @@
+		for (int i = 0; i < length; i++) {
+			Object[] arg = params[i];
+			Parameter<?> parameter = parameters[i];
+
+			if (arg == null || arg.length == 0) {
+				if (parameter.isOptional()) {
+					continue;
+				} else {
+					return null;
+				}
+			}
+
+			if (arg.length == 1 || parameter.isSingleValue()) {
+				assert parameter.getType().getC().isAssignableFrom(arg[0].getClass())
+					: "argument type %s does not match parameter type %s".formatted(parameter.getType().getC().getSimpleName(),
+					arg[0].getClass().getSimpleName());
+
+				args.put(parameter.getName(), arg[0]);
+			} else {
+				assert parameter.getType().getC().isAssignableFrom(arg.getClass())
+					: "argument type %s does not match parameter type %s".formatted(parameter.getType().getC().getSimpleName(),
+					arg.getClass().getSimpleName());
+
+				args.put(parameter.getName(), arg);
+			}
+		}
+
+		FunctionArguments arguments = new FunctionArguments(args);
+		T result = execute.apply(arguments);
+
+		if (result == null) {
+			return null;
+		} else if (result.getClass().isArray()) {
+			//noinspection unchecked
+			return (T[]) result;
+		} else {
+			//noinspection unchecked
+			T[] array = (T[]) Array.newInstance(result.getClass(), 1);
+			array[0] = result;
+			return array;
+		}
+	}
+
+	@Override
+	public boolean resetReturnValue() {
+		return true;
+	}
+
+	/**
+	 * Returns this function's description.
+	 *
+	 * @return The description.
+	 */
+	public @NotNull String[] description() {
+		return description;
+	}
+
+	/**
+	 * Returns this function's version history.
+	 *
+	 * @return The version history.
+	 */
+	public @NotNull String[] since() {
+		return since;
+	}
+
+	/**
+	 * Returns this function's examples.
+	 *
+	 * @return The examples.
+	 */
+	public @NotNull String[] examples() {
+		return examples;
+	}
+
+	/**
+	 * Returns this function's keywords.
+	 *
+	 * @return The keywords.
+	 */
+	public @NotNull String[] keywords() {
+		return keywords;
+	}
+
+	/**
+	 * Registers this function.
+	 *
+	 * @return This function.
+	 * @see Functions#register(DefaultFunction)
+	 */
+	@Contract(" -> this")
+	public DefaultFunction<T> register() {
+		Functions.register(this);
+
+		return this;
+	}
+
+	public static class Builder<T> {
+
+		private final String name;
+		private final Class<T> returnType;
+		private final Map<String, Parameter<?>> parameters = new LinkedHashMap<>();
+
+		private ch.njol.skript.util.Contract contract = null;
+
+		private String[] description, since, examples, keywords;
+
+		private Builder(@NotNull String name, @NotNull Class<T> returnType) {
+			Preconditions.checkNotNull(name, "name cannot be null");
+			Preconditions.checkNotNull(returnType, "return type cannot be null");
+
+			this.name = name;
+			this.returnType = returnType;
+		}
+
+		/**
+		 * Sets this function builder's {@link Contract}.
+		 *
+		 * @param contract The contract.
+		 * @return This builder.
+		 */
+		@Contract("_ -> this")
```

Comment:
we might actually want to say
```suggestion
		@Contract("null -> fail; _ -> this")
```
for these methods given the preconditions checks

Reply:
since the param is annotated as NotNull, IntellIJ gives a warning. should i still do it?

Reply:
ah okay it should be good as-is then